pkg/apis/certmanager/v1alpha1/types.go

```diff
@@ -88,8 +88,34 @@ type IssuerConfig struct {
 	// Certificate renew before expiration duration
 	RenewBefore time.Duration `json:"renewBefore,omitempty"`
 
-	ACME *ACMEIssuer `json:"acme,omitempty"`
-	CA   *CAIssuer   `json:"ca,omitempty"`
+	ACME  *ACMEIssuer  `json:"acme,omitempty"`
+	CA    *CAIssuer    `json:"ca,omitempty"`
+	Vault *VaultIssuer `json:"vault,omitempty""`
+}
+
+type VaultIssuer struct {
+	// Vault authentication
+	Auth VaultAuth `json:"auth"`
+	// Server is the vault connection address
+	Server string `json:"server"`
+	// Vault URL path to the certificate role
+	Path string `json:"path"`
+}
+
+// Vault authentication  can be configured:
+// - With a secret containing a token. Cert-manager is using this token as-is.
+// - With a secret containing a AppRole. This AppRole is used to authenticate to
+//   Vault and retrieve a token.
+type VaultAuth struct {
+	// This Secret contains the Vault token key
+	TokenSecretRef SecretKeySelector `json:"tokenSecretRef,omitempty"`
+	// This Secret contains a AppRole and Secret
+	AppRole VaultAppRole `json:"appRoleSecret,omitempty"`
```

Comment:
From my understanding of approle tokens, cert-manager will need to support renewing the token itself, as it'll be the only component that possesses the full token.

Perhaps I'm wrong 🤔

Reply:
The approle is like a username/password. It gives you a vault token that is used for all other API calls to Vault. For now the authentication is done each time but a kind of token cache and renewal could be but in place to skip the auth part each time a certificate is renewed. Not quite sure is it is worth it (performance vs number of certificates to renew). Although with a short certificate duration and many certificates defined it could add up. Definitively something worth looking at eventually.